rcl/test/rcl/test_wait.cpp

```diff
@@ -684,3 +685,97 @@ TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), wait_set_get_allocator
   EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
 }
+
+// Test wait set init failure cases using mocks
+TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), wait_set_failed_init) {
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  // nullptr failures
+  rcl_ret_t ret =
+    rcl_wait_set_init(nullptr, 1, 1, 1, 1, 1, 0, context_ptr, rcl_get_default_allocator());
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+  EXPECT_TRUE(rcl_error_is_set());
+  rcl_reset_error();
+
+  ret =
+    rcl_wait_set_init(&wait_set, 1, 1, 1, 1, 1, 0, nullptr, rcl_get_default_allocator());
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret);
+  EXPECT_TRUE(rcl_error_is_set());
+  rcl_reset_error();
+
+  {
+    // Mock rmw implementation to fail init
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rcl", rmw_create_wait_set, nullptr);
+    ret =
+      rcl_wait_set_init(&wait_set, 1, 1, 1, 1, 1, 0, context_ptr, rcl_get_default_allocator());
+    EXPECT_EQ(RCL_RET_WAIT_SET_INVALID, ret);
+    EXPECT_TRUE(rcl_error_is_set());
+    rcl_reset_error();
+  }
+}
+
+// Test failure init when using a bad allocators
+TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), wait_set_failed_init_bomb_alloc) {
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = RCL_RET_OK;
+
+  // Pass bomb allocator to make resize function fail
+  rcl_allocator_t bomb_alloc = get_time_bombed_allocator();
+  for (int i = 0; i < 10; i++) {
+    set_time_bombed_allocator_count(bomb_alloc, i);
+    ret =
+      rcl_wait_set_init(&wait_set, 1, 0, 0, 1, 1, 0, context_ptr, bomb_alloc);
+    if (RCL_RET_OK == ret) {
+      EXPECT_EQ(RCL_RET_OK, rcl_wait_set_fini(&wait_set));
+      break;
+    } else {
+      EXPECT_EQ(RCL_RET_BAD_ALLOC, ret);
+      rcl_reset_error();
+    }
+  }
+}
+
+// Test wait set fini failure cases using mocks
+TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), wait_set_failed_fini) {
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret =
+    rcl_wait_set_init(&wait_set, 1, 1, 1, 1, 1, 0, context_ptr, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret);
+  {
+    // Mock rmw implementation to fail fini
+    auto mock = mocking_utils::inject_on_return(
+      "lib:rcl", rmw_destroy_wait_set, RMW_RET_ERROR);
+    EXPECT_EQ(RCL_RET_WAIT_SET_INVALID, rcl_wait_set_fini(&wait_set));
+    EXPECT_TRUE(rcl_error_is_set());
+    rcl_reset_error();
+  }
+}
+
+// Test proper error handling with a fault injection test
+TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), wait_set_test_maybe_init_fail) {
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = RCL_RET_OK;
+  rcl_allocator_t alloc = rcl_get_default_allocator();
+
+  RCUTILS_FAULT_INJECTION_TEST(
+  {
+    ret = rcl_wait_set_init(&wait_set, 1, 0, 0, 1, 1, 0, context_ptr, alloc);
+    if (RCL_RET_OK == ret) {
+      EXPECT_EQ(RCL_RET_OK, rcl_wait_set_fini(&wait_set));
```

Comment:
@Blast545 what if `rcl_wait_set_fini` fails due to fault injection?

Reply:
Let me know if you think 6c2938c is good enough

Reply:
I think it should be enough, yes. `rmw_wait_set_fini()` keeps going on failure. 

I still think we have to revisit fault injection on `fini` functions, to avoid artificial leaks. But that's another discussion.